ols/src/cache/postgres_cache.py

```diff
@@ -0,0 +1,142 @@
+"""Cache that uses Postgres to store cached values."""
+
+import pickle
+from typing import Optional
+
+from langchain_core.messages.base import BaseMessage
+
+from ols.src.cache.cache import Cache
+
+
+class PostgresCache(Cache):
+    """Cache that uses Postgres to store cached values."""
+
+    CREATE_CACHE_TABLE = """
+        CREATE TABLE IF NOT EXISTS cache (
+            user_id         text NOT NULL,
+            conversation_id text NOT NULL,
+            key             text UNIQUE NOT NULL,
+            value           bytea,
+            updated_at      timestamp,
+            PRIMARY KEY(user_id, conversation_id)
+        );
+        """
+
+    CREATE_INDEX = """
+        CREATE INDEX IF NOT EXISTS timestamps
+            ON cache (updated_at)
+        """
+
+    SELECT_STATEMENT = """
+        SELECT value
+          FROM cache
+         WHERE user_id=%s and conversation_id=%s limit 1
+        """
+
+    UPDATE_STATEMENT = """
+        UPDATE cache
+           SET value=%s, updated_at=CURRENT_TIMESTAMP
+         WHERE user_id=%s AND conversation_id=%s
+        """
+
+    INSERT_STATEMENT = """
+        INSERT INTO cache(user_id, conversation_id, value, updated_at)
+        VALUES (%s, %s, %s, CURRENT_TIMESTAMP)
+        """
+
+    DELETE_STATEMENT = """
+        DELETE FROM cache
+         WHERE (user_id, conversation_id) in
+               (SELECT user_id, conversation_id FROM cache ORDER BY updated_at LIMIT
+        """
+
+    def __init__(self) -> None:
+        """Create a new instance of the `PostgresCache` class."""
+        import psycopg2
+
+        # initialize connection to DB
+        connection_string = "postgresql://localhost:5432"  # TODO from config
+        self.conn = psycopg2.connect(connection_string)
+        self.initialize_cache()
```

Comment:
Is it a good practice to execute in init?

Reply:
well why not? there are two outcomes:
1) the object is constructed correctly with all bits and pieces
2) it throws exception

nothing in-between

Reply:
Just thinking if this should happen when the object is initialized.
But I guess it is alright in this case. Otherwise, there would need to be some ensure it happens when the cache is actually used.

Btw should it print? Not close connection, log and raise?

Reply:
correct, it should close+log+re-raise (will do)

Reply:
fixed